src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/MapCode/MapCodeEndpoint.cs

```diff
@@ -0,0 +1,557 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.MapCode.Mappers;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Extensions;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+using SyntaxNode = Microsoft.AspNetCore.Razor.Language.Syntax.SyntaxNode;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.MapCode;
+
+/// <summary>
+/// Maps requested code to a given Razor document.
+/// </summary>
+/// <remarks>
+/// This class and its mapping heuristics will likely be constantly evolving as we receive
+/// more advanced inputs from the client.
+/// </remarks>
+[LanguageServerEndpoint(LSP.MapperMethods.WorkspaceMapCodeName)]
+internal sealed class MapCodeEndpoint : IRazorDocumentlessRequestHandler<LSP.MapCodeParams, LSP.WorkspaceEdit?>
+{
+    private readonly IRazorDocumentMappingService _documentMappingService;
+    private readonly DocumentContextFactory _documentContextFactory;
+    private readonly ClientNotifierServiceBase _languageServer;
+    private readonly FilePathService _filePathService;
+
+    public MapCodeEndpoint(
+        IRazorDocumentMappingService documentMappingService,
+        DocumentContextFactory documentContextFactory,
+        ClientNotifierServiceBase languageServer,
+        FilePathService filePathService)
+    {
+        _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+        _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+        _filePathService = filePathService ?? throw new ArgumentNullException(nameof(filePathService));
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public async Task<LSP.WorkspaceEdit?> HandleRequestAsync(
+        LSP.MapCodeParams request,
+        RazorRequestContext context,
+        CancellationToken cancellationToken)
+    {
+        // TO-DO: Apply updates to the workspace before doing mapping. This is currently
+        // unimplemented by the client, so we won't bother doing anything for now until
+        // we determine what kinds of updates the client will actually send us.
+        if (request.Updates is not null)
```

Comment:
Perhaps assert that this is null, so we know about it when/if the clientchanges?